File: 0127-word-ladder/0127-word-ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        int len = beginWord.length();
        unordered_map<string,vector<string>>allComboDict;
        for(string word:wordList){
            for(int i=0;i<len;i++){
                string newWord = word.substr(0,i)+"*"+word.substr(i+1,len);
                allComboDict[newWord].push_back(word);
            }
        }
        //queue for BFS
        queue<pair<string,int>>Q;
        Q.push({beginWord,1});
        //visited to make sure we dont repeat the processing same word
        unordered_map<string,bool>visited;
        visited[beginWord]=true;
        while(!Q.empty()){
            pair p = Q.front();
            Q.pop();
            string word=p.first;
            int level = p.second;
            for(int i=0;i<len;i++){
                //intermediate for current word
                string newWord = word.substr(0,i)+"*"+word.substr(i+1,len);
                for(string adjWord:allComboDict[newWord]){
                    if(adjWord==endWord){
                        return level+1;
                    }
                    if(!visited[adjWord]){
                        visited[adjWord]=true;
                        Q.push(make_pair(adjWord,level+1));
                    }
                }
            }
        }
        return 0;
    }
};

File: 0127-word-ladder/0127-word-ladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> wordSet = new HashSet<>(wordList);
        if (!wordSet.contains(endWord)) return 0;

        // Step 1: Preprocess - Build adjacency map using wildcard patterns
        Map<String, List<String>> patternMap = new HashMap<>();
        int wordLen = beginWord.length();

        for (String word : wordSet) {
            for (int i = 0; i < wordLen; i++) {
                String pattern = word.substring(0, i) + "*" + word.substring(i + 1);
                patternMap.computeIfAbsent(pattern, k -> new ArrayList<>()).add(word);
            }
        }

        // Step 2: BFS
        Queue<String> queue = new LinkedList<>();
        Set<String> visited = new HashSet<>();
        queue.offer(beginWord);
        visited.add(beginWord);

        int level = 1;

        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int j = 0; j < size; j++) {
                String current = queue.poll();

                for (int i = 0; i < wordLen; i++) {
                    String pattern = current.substring(0, i) + "*" + current.substring(i + 1);
                    List<String> neighbors = patternMap.getOrDefault(pattern, Collections.emptyList());

                    for (String neighbor : neighbors) {
                        if (neighbor.equals(endWord)) return level + 1;

                        if (!visited.contains(neighbor)) {
                            visited.add(neighbor);
                            queue.offer(neighbor);
                        }
                    }
                }
            }
            level++;
        }

        return 0;
    }
}
